dance/dance_pipeline.py

```diff
@@ -100,9 +169,236 @@ def _generate_molecules_from_database(self) -> oechem.OEMol:
                 ifs = oechem.oemolistream(mol2file)
                 mol = oechem.OEMol()
                 oechem.OEReadMolecule(ifs, mol)
                 yield mol
+                ifs.close()
+
+    def assign_fingerprint(self,
+                           fingerprint_function,
+                           output_oeb: Union[str, pathlib.Path] = "fingerprint_output.oeb"):
+        """Assigns a fingerprint to the molecules from the :meth:`~filter` step.
+
+        Each molecule is passed to the ``fingerprint_function`` to create a
+        fingerprint. This fingerprint is attached to the molecule, and the
+        molecule is saved to the ``output_oeb``.
+
+        Specifically, the fingerprints are stored as data entries in each
+        molecule and may be accessed later with ``OEMol.GetIntData()`` and
+        ``OEMol.GetDoubleData()``. The following data fields are stored in each
+        molecule:
+
+        - { :attr:`~FINGERPRINT_LENGTH_NAME` } (`int`) - the number of values in the
+          fingerprint
+        - { :attr:`~FINGERPRINT_VALUE_NAME` }_{i} (`float`) - the values of the
+          fingerprint, with i = [0,1,2,...,{ :attr:`~FINGERPRINT_LENGTH_NAME` } - 1]
+
+        Note that :attr:`~FINGERPRINT_LENGTH_NAME` and
+        :attr:`~FINGERPRINT_VALUE_NAME` are attributes of this class.
+
+        ``output_oeb`` is stored in the ``fingerprint_output_oeb`` attribute as
+        a ``pathlib.Path`` for future reference by the pipeline.
+
+        Parameters
+        ----------
+        fingerprint_function : function(molecule) -> array_like[float]
+            A function which provides the fingerprint for a single OEMol. This
+            fingerprint must be an array of floats.
+        output_oeb : Union[str, pathlib.Path], optional
+            Name of an OEB (Openeye Binary) file for storing the molecules along
+            with their fingerprints. If this file already exists, it will be
+            overwritten!
+        """
+        self.fingerprint_output_oeb = pathlib.Path(output_oeb)
+
+        filtered_molecule_stream = oechem.oemolistream(str(self.filter_output_oeb))
+        fingerprinted_molecule_stream = oechem.oemolostream(str(output_oeb))
+        for mol in filtered_molecule_stream.GetOEMols():
+            fingerprint = fingerprint_function(mol)
+            self._add_single_fingerprint(mol, fingerprint)
+            oechem.OEWriteMolecule(fingerprinted_molecule_stream, mol)
+        filtered_molecule_stream.close()
+        fingerprinted_molecule_stream.close()
+
+    def _add_single_fingerprint(self, mol: oechem.OEMol, fingerprint: "array-like of float"):
+        """Adds the fingerprint to the molecule in-place."""
+        mol.SetIntData(self.FINGERPRINT_LENGTH_NAME, len(fingerprint))
+        for i, val in enumerate(fingerprint):
+            mol.SetDoubleData(f"{self.FINGERPRINT_VALUE_NAME}_{i}", val)
+
+    def select(self,
```

Comment:
This is the core of the PR. To view the rendered version of the docstring, visit https://dance.readthedocs.io/en/btjanaka-selection/api/dance.DancePipeline.html#dance.DancePipeline.select